src/magic_folder/create.py

```diff
@@ -124,33 +41,68 @@ def magic_folder_create(alias, nickname, name, node_directory, local_dir, poll_i
     :param HTTPClient treq: An ``HTTPClient`` or similar object to use to make
         the queries.
 
-    :return Deferred[integer]: A status code of 0 for a successful execution. Otherwise
-        an appropriate exception is raised.
+    :return Deferred: ``None`` or an appropriate exception is raised.
     """
 
-    # make sure we don't already have a magic-folder with this name before we create the alias
-    maybe_upgrade_magic_folders(node_directory)
-    folders = load_magic_folders(node_directory)
+    # XXX probably want to pass this in, instead of "treq"?
+    tahoe_client = create_tahoe_client(config.tahoe_client_url, treq)
 
-    if name in folders:
+    if name in config.list_magic_folders():
         raise Exception("Already have a magic-folder named '{}'".format(name))
 
-    rc = yield tahoe_create_alias(node_directory, alias, treq)
-    if rc != 0:
-        raise Exception("Failed to create alias")
+    # create our author
+    author = create_local_author(author_name)
 
-    if nickname is not None:
-        # inviting itself as a client
-        from twisted.internet import reactor
-        treq = HTTPClient(Agent(reactor))
-
-        try:
-            invite_code = yield _invite(node_directory, alias, nickname, treq)
-        except Exception as e:
-            raise Exception("Failed to invite after create: {}".format(str(e)))
-
-        rc = _join(invite_code, node_directory, local_dir, name, poll_interval, nickname)
-        if rc != 0:
-            raise Exception("Failed to join after create")
+    # create an unlinked directory and get the dmd write-cap
+    try:
+        collective_write_cap = yield tahoe_client.create_mutable_directory()
+    except BadResponseCode as e:
+        raise RuntimeError(
+            "Error from '{}' while creating mutable directory: {}".format(
+                config.tahoe_client_url,
+                e
+            )
+        )
 
-    returnValue(0)
+    try:
+        personal_write_cap = yield tahoe_client.create_mutable_directory()
+    except BadResponseCode as e:
+        raise RuntimeError(
```

Comment:
okay, a specific exception sounds good -- but should it come from `tahoe_client` and just be passed on (that is, bare `.create_mutable_directory()`) or should we make (another) specific exception for failing to create directories for "personal writecap" and "collective writecap"?

(I see `TahoeClient` now returns a semi-generic `TahoeAPIError` ... but that also seems very unlikely to be useful to a real user...)

It seems like `TahoeClient` could probably provide exceptions useful to the user from its API calls (e.g. `FailedToCreateDirectory` or something that says "a mutable directory could not be created: ...more-details...").

Reply:
`tahoe_client` might be able to do better than `BadResponseCode` sometimes, yea.  Like, maybe there is a well-defined http response that represents "I couldn't talk to enough storage servers to make that happen".  Then it could raise an exception type that means specifically *that*.  For other cases maybe it's not clear what's going on ... a raw "500 Internal Server Error" might not have much more of an interpretation than `BadResponseCode`.  That's fine though, we can have a mix of highly meaningful stuff and less so.  We can also continue to improve this incrementally over time (perhaps also including Tahoe-LAFS changes to report errors with a more meaningful http response ... if tahoe-lafs chooses to keep its current HTTP API at all :wink: )

Reply:
I added (specific) exceptions to `TahoeClient`.